### PR TITLE
Dk/803 domain client hold

### LINKED ISSUE #803: Provide a way for analysts to put domains into clientHold
### Story

As an analyst
I need to put a domain in clientHold 
so that it's paused from resolving in the .gov zone.

This is for _domain objects_; the domain request must not be deleted.

### Acceptance Criteria

An analyst can put a domain on clientHold to prevent resolution.
An analyst can remove the clientHold status.

### Additional Context

Implementation Note: 
* This ticket doesn't include _actually setting a domain on clientHold_, which is an EPP function (see #857). It should, however, do the work necessary up to the point of 857.
* * Use an action button on Django Admin to trigger the function (clientHold). 

This issue can presume that the mechanism whereby an analyst is made aware about (and validates the authenticity of) a delete request occurs via some other process (which is ripe for future innovation).

~~Question: does this affect existing db setup?~~

### Issue Links

Blocks #804 
Depends on #964

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #803 

## Changes

- domain.py: added Ready and On Hold to Domain states
- domain.py: added transitions for placing and removing holds
- admin.py: added dictionary for ACTION_BUTTONS to DomainAdmin response_change; refactored
  response_change into multiple methods implementing submit of action buttons
- domain_change_form.html: added buttons for Place Hold and Remove Hold to Domain edit view; buttons conditionally
  display based on current state (appropriate button displays when in Ready or On Hold states)
- test_admin.py/common.py: added test for analyst placing and removing holds 

## Context for reviewers

Place hold and Remove hold buttons display conditionally for Domains.  'Place hold' displays when Domain is in Ready state.
'Remove hold' displays when Domain is in 'On hold' state.  Buttons (and associated function when submitting) work for analysts
and administrators.

## Setup

Since all states for Domain are not currently implemented in code, and transitions between Domain states don't yet exist,
it does not make sense to add to fixtures.  So, sharing below steps for a dev to pull the code and add transitions which
allow Unknown state to be changed to Ready, in order to interact with an example in the django admin interface.

In order to put a domain to the Ready state, pull the code and make the following temporary modifications:

models/domain.py:
Line 326, change to:
@transition(field="state", source=[State.ONHOLD, State.UNKNOWN], target=State.READY)

templates/admin/django/admin/domain_change_form.html:
Line 13, change to:
{% elif original.state == original.State.ONHOLD or original.state == original.State.UNKNOWN %}

Once these modifications are done, navigate to the admin, choose domains, and select a domain which is in Unknown state.
You should now be able to click the 'Remove hold' button.  Once this is clicked, this Domain will be in the Ready state.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [x] Checked keyboard navigability
- [x] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

Domain state of READY (has button to Place hold):

Domain State - READY

Domain state of ON HOLD (has button to Remove hold):

Domain State - onhold

Domain state of UNKNOWN (has no button for placing or removing client hold):

Domain State - unknown

### CODE REVIEW

**Comment by therealslimhsiehdy** [comment]
I think you may need to run a migration as you made some changes in `model/domain.py`! Here are the commands below:

`docker-compose exec app bash` -- gets you into bash
`./manage.py showmigrations` -- shows the current migrations that are finished, all should have [x]
`./manage.py makemigrations` -- makes the migration
`./manage.py showmigrations` -- now you should see the new/updated migration with a [ ]
`./manage.py migrate [folder name here, ie registrar]`
`./manage.py showmigrations` -- the migration changes should now have a [x] by it

Sometimes needed below: 
`docker-compose down`
`docker-compose up`

File: src/registrar/admin.py [hunk 1]
```diff
@@ -175,31 +175,57 @@ class DomainAdmin(ListHeaderAdmin):
...
+            "PLACE_HOLD": "_place_client_hold",
+            "REMOVE_HOLD": "_remove_client_hold",
+            "EDIT_DOMAIN": "_edit_domain",
+        }
+        if ACTION_BUTTONS["PLACE_HOLD"] in request.POST:
+            return self.do_place_client_hold(request, obj)
+        elif ACTION_BUTTONS["REMOVE_HOLD"] in request.POST:
+            return self.do_remove_client_hold(request, obj)
+        elif ACTION_BUTTONS["EDIT_DOMAIN"] in request.POST:
+            return self.do_edit_domain(request, obj)
```
Comment (zandercymatics) [inline src/registrar/admin.py:188]: (Optional but Nice to have) Theres a neat [trick](https://stackoverflow.com/questions/9168340/using-a-dictionary-to-select-function-to-execute) you can do in python where you can actually embed a function as an 'object' inside of a dictionary, as so:
`myDict = {
    "PLACE_HOLD": self.do_place_client_hold,
    "REMOVE_HOLD": self.do_remove_client_hold,
    ...
}
`
Then you can call it:
`myDict[name]()` or `myDict.get(name, lambda: 'Invalid')()`

Basically I think you should be able to do something akin to this, in sorta psuedocode:
```
if action is in request.post:
    return action_buttons.get(request.post (maybe?))()
```

This might be worth exploring. I think in one of Alysias draft PRS shes adding quite a few more buttons to this area so, since we will need to refactor either way, storing all that inside of

**Review by zandercymatics — COMMENTED** [review]
Looks good so far, will add some more to this later on

File: src/registrar/admin.py [hunk 2]
```diff
@@ -175,31 +175,57 @@ class DomainAdmin(ListHeaderAdmin):
...
         return super().response_change(request, obj)
 
+    def do_place_client_hold(self, request, obj):
```
Comment (zandercymatics) [inline src/registrar/admin.py:191]: I like these refactors, nice work

File: src/registrar/templates/django/admin/domain_change_form.html [hunk 3]
```diff
@@ -8,8 +8,12 @@
 
 {% block field_sets %}
     <div class="submit-row">
+        {% if original.state == original.State.READY %}
```
Comment (zandercymatics) [inline src/registrar/templates/django/admin/domain_change_form.html:11]: Since you mentioned it during the demo: I think this approach is fine here. I don't have any objections to this

File: src/registrar/tests/test_admin.py [hunk 4]
```diff
@@ -36,6 +37,57 @@
...
+class TestDomainAdmin(TestCase):
+    def setUp(self):
+        self.site = AdminSite()
+        self.factory = RequestFactory()
+        self.admin = DomainAdmin(model=Domain, admin_site=self.site)
+        self.client = Client(HTTP_HOST="localhost:8080")
+        self.superuser = create_superuser()
+        self.staffuser = create_user()
+
+    def test_place_and_remove_hold(self):
```
Comment (zandercymatics) [inline src/registrar/tests/test_admin.py:49]: This will need to be expanded once the epplib stuff gets integrated. I think the best approach to do this would be to create an empty test with a skip tag on it so that work can be added later... There are examples of this in test_models_domain.py, but for your case it might look like:
```  
    @skip("Waiting on epp lib to implement")
    def test_place_and_remove_hold_epp(self):
        raise
 ```

File: src/registrar/admin.py (hunk 1 above)
Comment (dave-kennedy-ecs) [reply src/registrar/admin.py]: @zandercymatics thank you for this suggestion.  i have refactored response_change to incorporate your comment

File: src/registrar/admin.py (hunk 2 above)
Comment (dave-kennedy-ecs) [reply src/registrar/admin.py:193]: thanks

File: src/registrar/tests/test_admin.py [hunk 5]
```diff
@@ -36,6 +37,57 @@
 logger = logging.getLogger(__name__)
 
 
+class TestDomainAdmin(TestCase):
+    def setUp(self):
+        self.site = AdminSite()
+        self.factory = RequestFactory()
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_admin.py]: thanks, done

**Review by zandercymatics — APPROVED** [review]
Approved as there is just those minor changes... From a functionality perspective this is working great. Great work on this ticket!

File: src/registrar/tests/common.py [hunk 6]
```diff
@@ -427,11 +427,19 @@ def create_superuser():
...
     p = "userpass"
-    return User.objects.create_user(
+    staffuser = User.objects.create_user(
         username="staffuser",
         email="user@example.com",
         password=p,
     )
+    staffuser.is_staff = True
+    staffuser.save()
+    return staffuser
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/tests/common.py]: made this change.  did not change the name of the function.  I agree it is confusing.  I don't know yet the contexts of the other tests.  One looks clearly like it means staff user, not sure about the other.  so i left function name as is

File: src/registrar/templates/django/admin/domain_change_form.html (hunk 3 above)
Comment (dave-kennedy-ecs) [reply src/registrar/templates/django/admin/domain_change_form.html:11]: thanks for looking at this

File: src/registrar/tests/test_admin.py (hunk 4 above)
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_admin.py:49]: added this placeholder
